server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -21,66 +21,140 @@
  * Then given a specific document, it can retrieve the corresponding fields from the document's source.
  */
 public class FieldFetcher {
+
     public static FieldFetcher create(SearchExecutionContext context,
-                                      SearchLookup searchLookup,
-                                      Collection<FieldAndFormat> fieldAndFormats) {
+        Collection<FieldAndFormat> fieldAndFormats) {
+        List<String> nestedMappingPaths = context.nestedMappings().stream().map(ObjectMapper::name).collect(Collectors.toList());
+        return create(context, fieldAndFormats, nestedMappingPaths);
+    }
+
+    static FieldFetcher create(SearchExecutionContext context,
+        Collection<FieldAndFormat> fieldAndFormats, List<String> availableNestedMappings) {
 
         // Using a LinkedHashMap so fields are returned in the order requested.
         // We won't formally guarantee this but but its good for readability of the response
         Map<String, FieldContext> fieldContexts = new LinkedHashMap<>();
         List<String> unmappedFetchPattern = new ArrayList<>();
         boolean includeUnmapped = false;
+        Set<String> matchingNestedFieldPaths = determineMatchingNestedFieldPaths(fieldAndFormats, availableNestedMappings);
 
         for (FieldAndFormat fieldAndFormat : fieldAndFormats) {
             String fieldPattern = fieldAndFormat.field;
             if (fieldAndFormat.includeUnmapped != null && fieldAndFormat.includeUnmapped) {
                 unmappedFetchPattern.add(fieldAndFormat.field);
                 includeUnmapped = true;
             }
-            String format = fieldAndFormat.format;
 
             Collection<String> concreteFields = context.simpleMatchToIndexNames(fieldPattern);
             for (String field : concreteFields) {
                 MappedFieldType ft = context.getFieldType(field);
                 if (ft == null || context.isMetadataField(field)) {
                     continue;
                 }
-                ValueFetcher valueFetcher = ft.valueFetcher(context, format);
-                fieldContexts.put(field, new FieldContext(field, valueFetcher));
+                // only add leaf value fetchers here if they are not inside a nested object, otherwise they are handled later
+                if (isNestedObjectSubfield(field, matchingNestedFieldPaths) == false) {
+                    ValueFetcher valueFetcher = ft.valueFetcher(context, fieldAndFormat.format);
+                    fieldContexts.put(field, new FieldContext(field, valueFetcher));
+                }
+            }
+
+            // check if pattern matches nested field
+            for (String nestedFieldPath : matchingNestedFieldPaths) {
+                List<String> nextAvailableMappings = new ArrayList<>(availableNestedMappings);
+                nextAvailableMappings.remove(nestedFieldPath);
+                FieldFetcher nestedSubFieldFetcher = FieldFetcher.create(context, fieldAndFormats, nextAvailableMappings);
```

Comment:
It seems we'll try to look up all the fields that were passed in, even if we know they won't match this nested document (because they don't have the right nested path)?

Reply:
You are right, but the patterns that won't match the subset of the source also won't create much additional work I believe. Thats why I opted to leave it and simply rely on filtering down the view on the source map to keep the logic simpler. But I will investigate to see if this can be simplified.

Reply:
This should be adressed by changes in 9fb1c2f